postgres/attempt.go

```diff
@@ -540,6 +562,59 @@ func (w *worker) requestAttemptsForSpec(req coordinate.AttemptRequest, spec *wor
 	return attempts, err
 }
 
+// chooseMakeFailAttempts finds work units to do for a specific work
+// spec, honoring the work spec's MaxRetries field.  It doesn't make
+// sense to call this with maxRetries as zero; call
+// chooseAndMakeAttempts instead.
+func (w *worker) chooseMakeFailAttempts(
+	tx *sql.Tx,
+	spec *workSpec,
+	numUnits int,
+	maxRetries int,
+	now time.Time,
+	length time.Duration,
+) ([]coordinate.Attempt, error) {
+	var attempts []coordinate.Attempt
+	for len(attempts) < numUnits {
+		moreAttempts, err := w.chooseAndMakeAttempts(
+			tx, spec, numUnits-len(attempts), now, length)
+		if err != nil {
+			return nil, err
+		}
+		if len(moreAttempts) == 0 {
+			return attempts, nil
+		}
+
+		// For each of the (new) attempts, count the number of
+		// existing attempts for the work unit and maybe fail it.
+		// (It might be nice to do this in a batch, maybe moving
+		// this logic inside chooseAndMakeAttempts?)
```

Comment:
My immediate feel is that doing it this way involves somewhat more network I/O than doing a batch lookup but should be about the same net time (we want the database to do an index-only counting aggregation on an indexed value for each work unit).  It possibly increases the latency to get more work units, which could have a secondary effect of increasing the number of get-more-work conflicts.

My other intuition here is that the max-retries setting only really matters for computationally intensive Go work units, since the subprocess-based diffeo/python worker framework does a better job of forcibly killing and failing overtime work units (and the policy of "automatically retry at timeout" makes more sense there).